File: game_plugins/hase_und_igel_2018/shared/sc/plugin2018/Advance.java
package sc.plugin2018;

import com.thoughtworks.xstream.annotations.XStreamAlias;
import com.thoughtworks.xstream.annotations.XStreamAsAttribute;
import sc.plugin2018.util.GameRuleLogic;
import sc.shared.InvalidMoveException;

/**
 * Ein Vorwärtszug, um spezifizierte Distanz. Verbrauchte Karotten werden mit k = (distance * (distance + 1)) / 2
 * berechnet (Gaußsche Summe)
 */
@XStreamAlias(value = "advance")
public class Advance extends Action{

  @XStreamAsAttribute
  private int distance;

  public Advance(int distance) {
    super();
    this.distance = distance;
  }

  public Advance(int distance, int order) {
    this.order = order;
    this.distance = distance;
  }

  @Override
  public void perform(GameState state) throws InvalidMoveException {
    if (GameRuleLogic.isValidToAdvance(state, this.distance)) {
      state.getCurrentPlayer().changeCarrotsBy(- GameRuleLogic.calculateCarrots(this.distance));
      state.getCurrentPlayer().setFieldIndex(state.getCurrentPlayer().getFieldIndex() + distance);
      if (state.getTypeAt(state.getCurrentPlayer().getFieldIndex()) == FieldType.HARE) {
        state.getCurrentPlayer().setMustPlayCard(true);
      }
      // Setze letzte Aktion
      state.setLastAction(this);
    } else {
      throw new InvalidMoveException("Vorwärtszug um " + this.distance + " Felder ist nicht möglich");
    }
  }

  /**
   * Gibt das Dinstanzattribut zurück
   * @return Distanz
   */
  public int getDistance() {
    return distance;
  }

  @Override
  public Advance clone() {
    return new Advance(this.distance, this.order);
  }

  @Override
  public boolean equals(Object o) {
    if(o instanceof Advance) {
      return (this.distance == ((Advance) o).distance);
    }
    return false;
  }

  @Override
  public String toString() {
    StringBuilder b = new StringBuilder("Advance: distance ");
    b.append(this.distance);
    b.append(" order ");
    b.append(this.order);
    return b.toString();
  }

}

File: plugin2025/src/main/kotlin/sc/plugin2025/Advance.kt
package sc.plugin2025

import com.thoughtworks.xstream.annotations.XStreamAlias
import com.thoughtworks.xstream.annotations.XStreamAsAttribute
import com.thoughtworks.xstream.annotations.XStreamImplicit
import sc.plugin2025.util.HuIConstants
import sc.shared.IMoveMistake

/**
 * Ein Vorwärtszug um die angegebene Distanz.
 * - verbrauchte Karotten werden mit k = (distance * (distance + 1)) / 2 berechnet (Gaußsche Summe).
 * - falls der Zug auf einem Hasenfeld endet, müssen auszuführende Hasenkarten mitgegeben werden.
 * - falls der Zug auf einem Marktfeld endet, wird eine zu kaufende Hasenkarte mitgegeben.
 *   Der Wert der Karottentauschkarte spielt dann keine Rolle.
 */
@XStreamAlias(value = "advance")
class Advance(
    @XStreamAsAttribute val distance: Int,
    @XStreamImplicit private vararg val cards: Card
): Move {
    
    val cost: Int
        get() = GameRuleLogic.calculateCarrots(distance)
    
    @Suppress("USELESS_ELVIS")
    fun getCards() = cards ?: arrayOf()
    
    override fun perform(state: GameState): IMoveMistake? {
        val player = state.currentPlayer
        val check = state.checkAdvance(distance)
        if(check != null)
            return check
        player.advanceBy(distance)
        
        var lastCard: Card? = null
        var bought = false
        return getCards().firstNotNullOfOrNull {
            if(bought)
                return HuIMoveMistake.MUST_BUY_ONE_CARD
            if(state.currentField == Field.MARKET) {
                return@firstNotNullOfOrNull player.consumeCarrots(10) ?: run {
                    bought = true
                    player.addCard(it)
                    null
                }
            }
            if(state.currentField != Field.HARE || lastCard?.moves == false)
                return HuIMoveMistake.CANNOT_PLAY_CARD
            lastCard = it
            it.perform(state)
        } ?: HuIMoveMistake.MUST_BUY_ONE_CARD.takeIf {
            // On Market field and no card bought or just moved there through card
            state.currentField == Field.MARKET && !bought
        } ?: HuIMoveMistake.MUST_PLAY_CARD.takeIf {
            // On Hare field and no card played or just moved there through card
            state.currentField == Field.HARE && lastCard?.moves != false
        }
    }
    
    @Suppress("SAFE_CALL_WILL_CHANGE_NULLABILITY", "UNNECESSARY_SAFE_CALL")
    override fun toString(): String =
        "Vorwärts um $distance${cards?.takeUnless { it.isEmpty() }?.joinToString(prefix = " mit Karten [", postfix = "]") { it.label } ?: ""}"
    
    override fun equals(other: Any?): Boolean {
        if(this === other) return true
        if(other !is Advance) return false
        if(distance != other.distance) return false
        return getCards().contentEquals(other.getCards())
    }
    
    override fun hashCode(): Int =
        distance + getCards().contentHashCode() * HuIConstants.NUM_FIELDS
    
}